misk/src/main/kotlin/misk/security/ssl/KeystoreExtensions.kt

```diff
@@ -0,0 +1,102 @@
+package misk.security.ssl
+
+import java.security.KeyStore
+import java.security.KeyStoreException
+import java.security.NoSuchAlgorithmException
+import java.security.PrivateKey
+import java.security.UnrecoverableKeyException
+import java.security.cert.X509Certificate
+import javax.net.ssl.TrustManagerFactory
+
+data class CertificateAndKey(val certificate: X509Certificate, val privateKey: PrivateKey)
+
+/**
+ * Read a certificate and key pair from the keystore.
+ *
+ * @param keyStore contains key pair.
+ * @param alias KeyStore entry alias.
+ * @param password passphrase protecting PrivateKeyEntry.
+ * @return KeyPair if found, Optional.absent() otherwise.
+ * @throws RuntimeException if passphrase is incorrect or keyStore uninitialized.
+ */
+fun KeyStore.getCertificateAndKey(alias: String, passphrase: CharArray): CertificateAndKey? {
+    try {
+        if (!entryInstanceOf(alias, KeyStore.PrivateKeyEntry::class.java)) return null
+        val certificate = getCertificate(alias) as X509Certificate
+        val key = getKey(alias, passphrase) as PrivateKey
```

Comment:
Where does `getKey` and `getCertificate` come from?

Reply:
Oh wow, didn't realize Kotlin supports extending classes this way. I see what's happening. Neat.